### PR TITLE
Merge of haskell-with-nixpkgs back into HNix

### LINKED ISSUE #593: Find a CI that is able to pass multiple check results into GitHub
Travis and most others CIs not even able to show multiple checks into GitHub.

CI should report to GitHub what builds are passed, good if show the short log/err message.

GHC platform and versions, warning checks, HLint, Haddock coverage... etc should be separate checks instead of ONE that CIs think is enough.

Even I myself used to configure private Hydra in 2017 to pass multiple checks into GitHub.

### LINKED ISSUE #653: Please, provide in CI the classical Cabal build
Long story short.

To see that something is the upstream Nixpkgs problem - classical Cacal CI build is needed so we are able to know when Nixpkgs builds have issues does classical Cabal+Hackage builds properly.

I already reached the Travis CI design constraints, since Travis CI has single `script:` process thread - Travis CI can not do two different build systems.

If we continue using Travis CI - it becomes fetters on the project automation growth, the level of heuristic complexity of its configuration would increase drastically, since its design forces to mesh-in and cross-dodge between different build processes (systems) inside a single `script:` thread. That would turn the configuration into unreadable, brittle code from currently elegant transparent code. Travis CI is just not designed for two different build systems.
Of course, Shell is Turing complete and project can offload everything into side-car code, but that means we start to write own CI platform, while CI itself required to do it in the first place.

And to make any migrations and keep with responsibilities of growing the project, the level of permissions should be raised to taken responsibilities and work, which I started to discuss in #636.

I kept our Travis CI setup working right with yesterday's sudden Nix installer release, and its changes and its current "caveats", and made a refactoring of configuration so it is in a good state now.

The community needs to be able to monitor, be helped to understand the Nixpkgs inconsistency problems when they arise, we need CI extension/migrations.

I want and it is seen - took responsibility to work, extending our processes and automation for everyone to have a smooth going contribution experience, I need to be given permissions to proceed.

---

I am a proponent of the sticking to Nixpkgs as much as possible for the project, but I would build a solid case of contra Nixpkgs here to be realistic:

Prelude: Currently, CI uses Nixpkgs store for all builds, 5 different GHC versions and so their Nix closures `{GHCJS, 8.4.4, 8.6.5, 8.8.3, 8.10.1}`.

1. By current Nixpkgs design, the Nixpkgs really consolidated over particular GHC that is the default in it, currently, it is `8.8.3`.
  This design towards one version - results in less use, maintenance, eyes and hands on other GHC versions and so all their store branch consistency in Nixpkgs, and source structure and community treating them as additional functionality.
2. Nixpkgs upstream practice to dodge upstream progress with creating side-packages for new releases - results into that Nixpkgs constantly seduces itself into dodging the Hackage progress, and so downstream gets inconsistency of newest GHC version subset in Nixpkgs.
3. Lack of upstream automation results into lack of hands for upstream manual maintenance.
4. {3,1,2} would constantly resolve into that graph trees of non-default in Nixpkgs GHC versions would be constantly sacrificed to keep default Nixpkgs GHC graph consistent, community forgets to account for all GHCs, default GHC would always get a maintainers priority, and dodging upstream and overall side-packages practice allows to keep default GHC graph working on the default packages consistent, while breaks all other GHCs graphs consistency, and so those branches now need switch/make side-packages and so they need overrides of packages to side-packages. And as stated overall the consistency of non-default Nixpkgs GHCs is less then ideal.

---

Good thing - we are the very specialists that can debug, work and automate Nixpkgs, specialists that in time can change/improve something in Nixpkgs solely by making HNix a reality.

---

This Nixpkgs situation results in that the project frequently hits the Nixpkgs upstream package version inconsistencies of the store, like companion got in https://github.com/haskell-nix/hnix/pull/635#issuecomment-648726295

An example of nearest empirical proof of the situation and statement is seen in: https://travis-ci.org/github/haskell-nix/hnix/builds/701250357, due to need of one current Hackage version - all GHC graphs got broken except the default GHC due to current Nixpkgs design. Good that Nixpkgs is mostly the freshest store around.

### LINKED ISSUE #683: Please, update Nixpkgs GHC to 8.8.4
Yesterday new default GHC arrived into Nixpkgs, which requires `default.nix` and build infrastructure to follow.

Due to complex merge in the process, I bundled this update into https://github.com/haskell-nix/hnix/pull/681.

### LINKED ISSUE #684: Please, remove from CIs the use of Nixpkgs GHC 8.4.4, since it is deprecated
Nixpkgs keeps a promise to deliver last three major versions of GHCs, and 8.4 is currently fourth.

After yesterday's updates in Nixpkgs the GHC 8.4.4 no longer present normally.

It is still present in many Nixpkgs files, but:

```fish
nix-env -f "<nixpkgs>" -qaP -A haskell.compiler | awk '{print $2}' | sort | uniq

ghc-8.10.1
ghc-8.11.20200505
ghc-8.2.2-binary
ghc-8.6.5
ghc-8.6.5-binary
ghc-8.8.2
ghc-8.8.3
ghc-8.8.4
ghcjs-8.6.0.1
```

---

Due to complex merge in the process, I bundled this update into https://github.com/haskell-nix/hnix/pull/681.

### PR DESCRIPTION
**Pull request by Anton-Latukha**
This PR update build processes, both with CI processes:
  *  Migrates project to the GitHub Actions CI
    * :heavy_check_mark: updates from Travis 50min restriction to GitHub 6 hours of execution time, which allows churning through Nixpkgs revisions that when Hydra CI Cache is not present (since Hydra CI cache has a pretty constant lag to after the channel update).
    * :heavy_check_mark: faster node startup times
    * :heavy_check_mark: much bigger number of nodes and the compute power of nodes
    * :heavy_check_mark: Cachix and Nix deployment processes are better supported for GitHub Actions.
    * :heavy_check_mark: multiple checks reported to PR, so the community can roughly see the info on build results even without going into CI interface.  #593
  * :heavy_check_mark: expands the number of builds: https://github.com/haskell-nix/hnix/pull/681/checks
  * :heavy_check_mark: allows to better customize and tune the build processes
  * :heavy_check_mark: introduces classical Cabal Hackage builds (`8.{4,10}`), that is mandatory for the Haskell project, #653
  * CI updates:
    * :heavy_check_mark: Nixpkgs GHC 8.8.(3->4), #683 
    * :heavy_check_mark: Nixpkgs drop of GHC 8.4.4, migration 8.4.4 to Hackage build, #684
  * etc...

Also:
  *  :heavy_check_mark:  While waiting on GitHub CI Cachix key setup (#682), reduced Travis, all builds become optional and added "bypass build" for fast skip, it is currently ours "green" background Cachix builder.

---

`HNix` was taken and with maximal preservation of Git history, Git history was
passed through pass filter and that in how `haskell-with-nix` was formed.

Further, work on CI and build configs happened there.

Due to `HNix` being a mothership project and since `haskell-with-nix`
preservation of original history - it is possible to merge progress from it
back.

Now, when `haskell-with-nixpkgs` main setup stabilized - it is ready to be
merged back, meanwhile `HNix` needed to dodge Nixpkgs downstream quirks which
blocked the development processes and new CI was needed.

Merge went pretty ideal, for merge `haskell-with-nixpkgs` only relevant history
was also filtered-out and also dropped 3 commits that got into the passed files.

A custom merge commit also was required.

But overall, the process of merge went unexpectedly great.

---

After the inside PR merge - also the change of stub names used in
`haskell-with-nixpkgs` to `HNix`-specific names is obviously needed. So that directly follows this
merge inside PR branch.

### CODE REVIEW

**Comment by Anton-Latukha** [comment]
@jwiegley 

To use Cachix upload in GitHub CI - HNix needs to provide its Cachix secretKey inside GitHub repository, the same key which is used in the current Travis setup.

The secret key should be stored by you somewhere, Cachix does not provide secret key restoration.

Please, put secret into the repository ([straightforward section of official docs](https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#creating-encrypted-secrets-for-a-repository)), please name secret as `CACHIX_SIGNING_KEY`, so we use the same name for the key everywhere, it is the variable name that Cachix expects and so also name wich env interfaces built around it await.

Meanwhile, at the same time waiting for this - I would tailor the merged template config to HNix-specific values.

**Comment by Anton-Latukha** [comment]
Currently, the tailoring of `hashell-with-nixpgs` template to `HNix`-specific names would happen.

**Comment by Anton-Latukha** [comment]
Here we go.

It works and looks normal, seems ready to be used.

File: default.nix
Comment (sjakobi) [inline default.nix:216]: Mustn't we assume that `neat-interpolation_0_5_1_1` will soon be gone too, when the next version of `neat-interpolation` is released or the latest version becomes the default?

File: default.nix
Comment (sjakobi) [inline default.nix:9]: I don't understand the motivation for having this parameter. Could you give an example where it would be useful?

File: .travis.yml
Comment (sjakobi) [inline .travis.yml:22]: Why do we still have the Travis config? Isn't GitHub CI sufficient?

File: .github/workflows/Cabal-Linux.yml
Comment (sjakobi) [inline .github/workflows/Cabal-Linux.yml:41]: What does `build20` mean? If it's just a name, how about referencing the main property/properties of this job?, e.g. `ghc804`.

**Review by sjakobi — COMMENTED** [review]
Wow! Looks like a massive amount of work! Will this help make CI more reliable, i.e. more robust against Nixpkgs changes and build caching issues?

Also, why is the GitHub CI config spread over 5 files? I assume that there must be a lot of duplication that might be tricky to keep in sync.

File: default.nix
Comment (sjakobi) [reply default.nix:216]: (That's why I had used `callPackage` in https://github.com/haskell-nix/hnix/pull/680)

File: default.nix
Comment (Anton-Latukha) [reply default.nix:9]: It is a ground work towards the support of monorepo with several projects.

Since `develoPackage` should include `name` of the Cabal file, and different projects have different Cabal names, passing through the `cabalName` and `projectRoot` - allows to traverse over the projects in monorepo. `hnix-store` is example of such Nix infrastructure - it has one `default.nix` for two projects.

Since HNix plans to move to monorepo at some near point - I preserved this ground work, there is also CI builds which matrix traverse over project roots `[ ./ ]`, aka singleton currently.

File: .travis.yml
Comment (Anton-Latukha) [reply .travis.yml:22]: Yep, still need it.

For a Nixpkgs cache loophole, explained in: https://github.com/haskell-nix/hnix/issues/682#issuecomment-667540898

Travis currently builds and pushes Cachix Nixpkgs cache that GitHub CI uses, while secret is not provided.

File: .github/workflows/Cabal-Linux.yml
Comment (Anton-Latukha) [reply .github/workflows/Cabal-Linux.yml:41]: Well, there is `name` field for it. ID of the builds and `name` are mandatory, and essentially fields that are duplicates, but required by the GitHub CI design.

GitHub CI has number of design choises I am complettely puzzled about. Like they not follow basic YAML RFC and so metadata build IDs strings are tightly restricted set of symbols, and inside there is the separate `name` field which follows YAML RFC, if they respected YAML RFC in IDs - that would've been one field.

There is `name` field, there is already manual tracking of renaming needed, since there already a `name` - metadata IDs can not bear any useful information only try to duplicate the `name` all the time, and because one has more important stuff then to duplicate sync same information translating it into ID required restricted form - made a choise for the `name` to shows the real human readable name and required for machines ID field is just old-style unstractured-programming type numbering that allows to put 9 builds in between them, since IDs does not really metter.

File: default.nix
Comment (Anton-Latukha) [reply default.nix:216]: If someone knows how we can circumvent this Nixpkgs flaw and always find import from the Nixpkgs the last release version of the package - I would be glad to know also.

Comment (Anton-Latukha) [reply default.nix:216]: The argument is to prefer Nixpkgs packages over custom overrides, because we test Nixpkgs builds, and if we override everything - that is not Nixpkgs test, that is custom Nix config test that has almost nothing with Nixpkgs.

Looking from the other side on the same argument:
Lets assume the contrary. Preferring override of the package to what is available in the store essentially means to prefer avoiding the global store for what it was created and prefer keeping custom store, then the question is why we even test for Nixpkgs then if we had chosen to avoid it.

So if we stick to Nixpkgs - we should try to follow it and be in touch in the processes in it. We may even have it as optional, but CI builds should track the state there.

Comment (Anton-Latukha) [reply default.nix:216]: Well, on the other side - I also do not like that the Nixpkgs quirks constantly break the CI pass through process.

As I said - lets make Nixpkgs optional. - That seems the most effective solution to the problem.

And after that we may think to compose some middle ground, we can have a build option with frozen Nixpkgs overrides, so if devs/users can not build with current Nixpkgs - README can direct them to use frozen version. Of course, we can just direct or expect them switch to other GHC sets if possible - that solves a part of that probabilistic issue, and of course we would need to maintain the frozen overrides if we provide them.

So currently making more Nixpkgs build optional seems the most effective solution.

Comment (sjakobi) [reply default.nix:216]: If we can make the Nixpkgs build optional, we might as well get rid of it.

I thought the basic problem was simply that the Nixpkgs revision is unpinned, and therefore each build might use a different and often incompatible configuration. I've never understood why we don't simply pin the revision and update it every few months or as needed.

File: default.nix
Comment (sjakobi) [reply default.nix:9]: > Since HNix plans to move to monorepo at some near point

Who is planning this? Is there an issue where this plan is discussed? If not, we should have one.

File: .travis.yml
Comment (sjakobi) [reply .travis.yml:22]: I'm not thrilled to babysit _two_ CI systems! Please get this fixed soon.

File: .github/workflows/Cabal-Linux.yml
Comment (sjakobi) [reply .github/workflows/Cabal-Linux.yml:41]: Thanks for the explanation!

**Comment by sjakobi** [comment]
> Also, why is the GitHub CI config spread over 5 files? I assume that there must be a lot of duplication that might be tricky to keep in sync.

Maybe using https://github.com/vmchale/github-actions-dhall would help. …only if it can actually reduce the complexity of the configs though…

File: default.nix
Comment (Anton-Latukha) [reply default.nix:216]: But the meaning of Nixpkgs build is Nixpkgs build tracking. Otherwise - we can just `cabal2nix` into custom Nix file and forget about Nixpkgs at all.

The argument that optional build can be reduced, it is analog to saying: since monitoring is optional - we can remove monitoring, since documentation is optional for operations - we can remove documentation.

Comment (Anton-Latukha) [reply default.nix:216]: The reason for Nixpkgs to be optional - is that since it is brittle - we would see the changes that needs to be made, but CI would pass PRs at the same time, and we can see that Cabal & Hackage & testing suites work and that error relates to downstream Nipkgs issue and not to the PR code, and so would remind us to fix it.

The idea is so we shift from sequence requirement `(code merge request with CI block due to Nixpkgs -> wait -> develop & merge Nixpkgs fix PR -> wait -> rebase CI tests & code merge -> HNix)` develop and fix Nixpkgs in separate simple processes and in parrallel `(code merge request -> HNix),  (PR on Nixpkgs fix -> HNix)` since that can be done in parallel.

Comment (Anton-Latukha) [reply default.nix:216]: As I said, using of Nix for Nixpkgs testing is essentially indeed optional, it is HNix being a good citizen for Nix ecosystem. But removing the tests - is taking a revolver and becoming Nixpkgs Wild West outlaw, that feeds on Nix ecosystem but has no regard to its laws and we would be like:
```
"

- (Nix community)  You broke everything. it is broken for 3 month now.

- (HNix)  HNix does not work and breaks stuff in Nixpkgs? - honestly, we do not care to monitor it, we removed test builds for it, we do not care about it to such degree.

- (Nix community)  Wait what is the agenda of the project?

- (HNix)  To support and evaluate Nix ecosystem, most importantly Nixpkgs.

- (Nix community)   ... wait, the agenda of the project is to work with and within Nix ecosystem - right? And you chosen to not care about Nixpkgs?.. Isn't that a direct collision that you disregard the purpose of the project and the community it was did for?

- (HNix)  ... Look. Here is the revolver.

"
```

If we decide to stop even monitoring Nixpkgs, we as well can just turn around and leave the project now.

**Comment by Anton-Latukha** [comment]
#### Workflows are the base of GitHub CI organization of config and processes

Because there is this knowledge of the workflows for the GitHub CI.

Look into:
  * https://github.com/haskell-nix/hnix/pull/681/checks
  * at the grouping of builds and at name of builds below

It is starting examples. GitHub workflows allow to logically construct the CI processes and much more, chain them, connect services as workflows and start them after some workflows passed, etc...

What is the benefit of using one workflow for all things? Everything is in one pile and sharing one combined env despite of the completely different build systems and conditions. And all being in one category.

Sometimes the configuration of systems is not like Haskell which is polymorphic and can deduplicate almost everything. Configuration languages are often very rigid, sometimes are limited and designed to be explicit and span configuration files with some minor explicit general conf shared in them between builds. Configuration of Travis and GitHub CI is a good example, where more explicitness helps and required. GitHub CI design demands this explicitness.

I've tried to build Travis CI matrix - it worked but customizing separate builds and naming them was impossible, not having customization abilities is not an effective strategy in CI. Essentially then one needs to multiply all parameters into gian matrix and then deprecate all instances that are not needed - result is - more complex configuration that requires growing a matrix and as a result growing a reverse logic to remove most of the matrix to live only needed builds.

I've could put everything in one file and build matrixes in GitHub CI also, then we would have the same problems - impossible to customize the builds and name custom builds properly, custom builds mean separate builds in these configs.

And actually separating CI config by logical workflows - makes it much nicer and cleaner, from my perspective.

File: default.nix
Comment (Anton-Latukha) [reply default.nix:216]: Regarding why we do not pin the Nixpkgs - the same question is "Why we do not freeze the Cabal configuration for Hackage."

Because there is no sense - if we pin Nixpkgs for our builds, then (very quickly) - the HNix would not work in new Nixpkgs, and then what is the sense in pinning Nixpkgs if all community constantly rolls on Nixpkgs and HNix does not work for them, so when we pin - we slash the community and the growth speed of it.

HNix can pin Nixpkgs to achieve interpreting Nixpgs successfully - then we internally stabilize the state so we can develop for it, but then we do not track interpretation of current Nixpkgs and would need to adjust to it afterward constantly anyway, since that is the main aim of the project.

1. So pinning for Nixpkgs to build HNix - results into us not supporting and not advertising project to Nixpkgs community which is where main community located.

2. And pinning Nixpkgs for interpretation in HNix - means project most probably does not work for current Nixpkgs and we, nor our users can not open Reports in Nixpkgs upstream, since HNix does not use current Nixpkgs but some old frozen one, and the first question everyone asks on debugging is "What version you are using, please check for the latest release and report only for the most recent release.", and since we are who pinned Nixpkgs internally - we are responsible for solving that and tracking latest Nixpkgs release, and we anyway would need to follow Nixpkgs always, since it is a technical target of the project.

File: .travis.yml
Comment (Anton-Latukha) [reply .travis.yml:22]: If someone should not be thrilled to babysit _two_ CI systems - is I. Since I did major work working on them and maintaining them and obvious that I am who would maintain them in the nearest future.

And would say even more. Since it is free tiers and our CIs in the future would have different paradigms (classical and FP, aka Hydra or Hercules or Sorkis CI) -  more CI systems - may be beneficial. Some CI would have good Windows testing, some would provide true FP Nix CI with all its abilities, some other services would provide supplementary functions.

I feel vibes that you may need to rest and relax. When we feel something or many things as important - often we would need to go on a walk and think about what is really important in life. I am the one that need to worry about CIs, not you. I am calm and reserved about all this, I just try to understand your argumentation, so I respond to you in the talk form, so we write back and forth to get understanding.

I gave you a strong argument why, Travis currently provides the cache for GitHub CI, until John puts the a secret key that only he holds and should hold. So currently there is Travis to produce the Nixpkgs cache that gets shared over Cachix so the GitHub CI does not rebuilds the Nixpkgs stack for 3 hours at every commit someone does in PR. It can not be fixed sooner then John appears and provides the secret key for us, until then all our Nixpkgs builds tied to Travis CI.

---

For example, see the response of Gabriel: https://github.com/dhall-lang/dhall-haskell/issues/1678#issuecomment-592960057

Sometimes 3-4 CIs are the best stable setup that Free Software Haskell project can do currently.

File: default.nix
Comment (Anton-Latukha) [reply default.nix:9]: It is `"Split parser & types out into different packages"` #328, there the mororepo approach was discussed much, several people from their experience voted up for it, John also voted up that he ready moreover then monorepo - even completely split the lib from executable, what is to do - is of course to what is most effective, what most effective should be planned.

The backtracking `haskell-with-nixpkgs` monorepo support back for merge into HNix - overcomplicated the merge process. The monorepo preparation here is minimal, it is several lines of code in `default.nix` and in GitHub workflows, and it is explained in the comments.

The monorepo code is easy and minimal, and to remove monorepo preparation is several easy commits, so I left it, since it is done, documented, awaited in this project, and always can be easily removed.

I've looked into `hnix-store` and thought if we would go monorepo - most probably it would be how it is done in `hnix-store`.

File: .travis.yml
Comment (Anton-Latukha) [reply .travis.yml:22]: Since Travis CI is supplementary at this point - I would make it optional and its check a fast skip, until we have normal cache process, so we would need to wait only on GitHub CI to pass, and Travis would always be green while would build cache at the moment.

Comment (sjakobi) [reply .travis.yml:22]: Sorry, @Anton-Latukha. It's hard for me stay relaxed about this when I receive an email from Travis CI informing me about the broken `master` branch every single day. Situations like the one in https://github.com/haskell-nix/hnix/pull/680 are also very frustrating to me because the CI failure appears to be completely unrelated to my work and there is no information on how the build could be fixed.

I very much hope that you will put an end to this misery.

File: default.nix
Comment (sjakobi) [reply default.nix:9]: Thanks for the link. I think it would be good to have a separate issue for the mono-repo discussion. It's currently rather hidden in a very long thread.

File: default.nix
Comment (sjakobi) [reply default.nix:216]: OK. I'm somewhat skeptical about the need to build with the latest Nixpkgs all the time.

My current understanding is that the Haskell packages in Nixpkgs are basically just downstream from Stackage. So it seems to me that we can simply try to keep Stackage happy, and let the Nixpkgs maintainers worry about the rest.

If you still want this monitoring, that's fine by me, but please ensure that it doesn't affect CI results and that new contributors can get a clean build with the Nix setup at all times.

Comment (Anton-Latukha) [reply default.nix:216]: How Nixpkgs can be downstream from the Stackage, if Nixpkgs autoimports downstream from the Hackage.

For example, HNix is on Hacakge, and not in Stackage, but in Nixpkgs, how it gets into Nixpkgs after we release it on Hackage, in the night it already appears in Nixpkgs.

`stack2nix` was working half way half a time, and it is recenly deprecated by creators, `stackage2nix` also abandoned project.

The `peti` build projects `hackage2nix` `cabal2nix`, Nixpkgs machinery uses them, process tied on regeneration of `configuration-hackage2nix.yaml` that is the entry point for auto imported Haskell Nixpkgs project descriptions. He has presentations on it, and indeed every 3:30 in the night autoimported Haskell packages in Nixpkgs sync to Hackage, that is why HNix updates in there and we see if it builds or there is breakage.

---

The thing is. Nix is always pain. FP OS management for many years ahead would be a pain, it would be pain until FP OS ecosystems mature (and it would be longer process then Haskell maturity, FP OS management is much harder ad-hock question and volume of particulars then the design of the programming language), it would be a pain until IT would not check to comply with it, I mean that their projects build in Nix or Guix. All hidden under the carpet troubles of IT would and ignorance to collaborate of its people, while hidden in classical approaches - would lay explicit marks on the newborn sick FP OS management child.

And as I stated, FP IT management is a much more complex topit - too much projects, project options, OS management (bootloaders, system level (all of systemd), all filesystems), build systems, languages, own language, own interpreter/compiler, own store, hardware particulars - there are too many questions and too many work to maintain all of it while FP OS management is so demanding for its quality - it would be in the process of maturing for a long time. There was a lot of scientific work for FP, but almost none for how to design FP IT management language and how to do FP IT management of all things.

It is and would be a constant PITA for a very long time, years and years. What I hope is that with migration on Haskell things would become bearable and velocity would have bigger acceleration, so much more exponentially rapid process would happen.

File: .travis.yml
Comment (sjakobi) [reply .travis.yml:22]: > I gave you a strong argument why, Travis currently provides the cache for GitHub CI, until John puts the a secret key that only he holds and should hold. So currently there is Travis to produce the Nixpkgs cache that gets shared over Cachix so the GitHub CI does not rebuilds the Nixpkgs stack for 3 hours at every commit someone does in PR. It can not be fixed sooner then John appears and provides the secret key for us, until then all our Nixpkgs builds tied to Travis CI.

Ping @jwiegley.

File: default.nix
Comment (sjakobi) [reply default.nix:216]: > For example, HNix is on Hacakge, and not in Stackage, but in Nixpkgs, how it gets into Nixpkgs after we release it on Hackage, in the night it already appears in Nixpkgs.

Good question. I don't know. See https://github.com/haskell-nix/hnix/issues/676#issue-656518706 though, specifically

> The problem is, that we get an older neat-interpolation ==0.3.2.6 from stack.

File: .travis.yml
Comment (Anton-Latukha) [reply .travis.yml:22]: Well, no hurry, I already told him a couple of times before, pinged him a couple of times, and created appropriate Report #682, it is not urgent. I've also could of course directly require this action in advance.

I've just made all Travis builds "optional" and enabled "fast skip" of them, and added a "bypass build" that returns "exit" as soon as instance boots up, so Travis always green and passes in 40 seconds for us, and builds and provides caches in the background.

**Comment by Anton-Latukha** [comment]
#### Turned off GHCJS build

∵:
1. GHCJS build fails for months and nobody showed up to care and pursue it.
2. GHCJS build makes global CI report status ":heavy_multiplication_x:", and there is GitHub CI option to pass CI further and so also PR, but there is no way for GitHub CI to not propagate that ":heavy_multiplication_x:" to global marker.
3. Compilation of Haskell to JS was a half measure until there would be a better solution, since it is much more logical to compile Haskell to WebAssembly, and there is https://github.com/tweag/asterius for that.

⊢ GHCJS build test turned off.

If someone would actively pursue GHCJS
  then we can enable it back
  else there is no meaning to keep building and testing and looking at it

**Comment by Anton-Latukha** [comment]
About using 'DHall' and 'github-actions-dhall' for CI (answering message https://github.com/haskell-nix/hnix/pull/681#issuecomment-667560666), lets see further down the line.

I do no see the problem of having several files where everything is structured and laid-out in official GitHub way without use of additional utilites, and vast majority of people would never look or care about CI configuration further down the line, it is important but supplementary to the main goal - the project itself.

But when the people would look into GitHub CI configuration - they would work with the standard GitHub CI setup configuration.

File: default.nix
Comment (Anton-Latukha) [reply default.nix:216]: That means that I need to learn more about Haskell Nix stack.

There is indeed Stack hooks.

Maybe part of the answer is that Stackage essentially is a Hackage, at least in terms of API, so same Hackage tools can download from both places.

Or most probably it is even simpler. Tool to get the consistent package set - receives a last LTS package version numbers from Stackage, and integrates those packages directly from Stackage or Hackage.

The full chain of `Hackage -> ... -> Nixpkgs install` is something I personally never thoroughly went through.

And actually, that is what I want to do, in previous talks I was talking to have a real Nixpkgs integration test, where the chain of integration would be a great representation of Hackage -> Nixpkgs integration build and test process, it would revolve around `hackage2nix`, `cabal2nix` and regeneration of `configuration-hackage2nix.yml`. So the integration test would closely replicate the process, that would sum up this our conversation.

This is my next agenda regarding the CI theme.

**Review by sjakobi — APPROVED** [review]
LGTM based on my impression that this addresses some of our current CI woes.

If there are any issues we'll probably detect them while using the CI in anger.

Adding hundreds of commits to `master` for this CI overhaul seems a little disproportionate, so maybe this PR should be squashed.

Thanks for all the work, @Anton-Latukha, and thanks for addressing my concerns with the current CI setup! :+1:

**Comment by Anton-Latukha** [comment]
Well, my view of it:

* commits should be small and atomic, note the steps of development, and so - be frequent and so have a footprint of development. And they are here. I just practice committing and tracking of the development.
* Every commit does a particular step, and action, since development is evolution, and the development footprint is there.
* Small commits are self-evident, and commit name often documentation of what and why was done.

* As I shown - Merge commits are abstractions over merged commits. In future I see only the workflows based on Merge abstractions.
  For example: I can take 670 commits, and organize them into merges by 10, then those by 10, we would get 1 main abstraction 7 sublevel abstractions, 70 subsubabstractions, 670 commits.

Specifically sitting and faking, retrofitting Git history is silly.
If some error or organization is on me - I always reorganize things on the spot locally.

---

The argument that there is some magic number of commits person needs to get into, while what matters is work done, and that the number of some commits and my number of commits different - is an argument of numbers, who likes which numbers.

While what primary - is results, and following a good practice.

And googling `git best practices`:
1. Git Tower: https://www.git-tower.com/learn/git/ebook/en/command-line/appendix/best-practices
2. https://deepsource.io/blog/git-best-practices/
3. http://sethrobertson.github.io/GitBestPractices/

---

Obviously I node towards this:

1. Small atomic commits is good
2. Small commits are self-documenting
3. Small commits mean great Git blaming
4. Small commits mean that during debugging (bisecting) person arrives directly towards the goal
5. Merge commits are perfect abstractions
6. Using merge commit abstractions with small commits inside them results in a more effective Git bisect algorithm.

What I mean, is that Git workflows can become much better.

**Comment by Anton-Latukha** [comment]
Thank you @sjakobi for the review.

---

I understand that my style sometimes may be seen as obnoxious.

I squashed everything into one commit.

My style honestly started from learning Git by doing, and Git was painful for me back in a dey, I used to not commit changes in time and created a small number of huge commits, so I started frequently commit unital atomic commits to get accustomed to Git commiting.

And, obviously, big commit numbers inspire respect both in person, and more importantly in the projects. And self-documentation, the following of the guidelines and developing effective practices - are a bonus.

If something - I encourage adoption of this style, since it arrives at many goals.

---

Thank you again, I am kneeling down for your work 🙏.

**Comment by sjakobi** [comment]
No worries, @Anton-Latukha! I appreciate your work on our CI! :)

**Pull request merged**
